pkg/util/push/request.go

```diff
@@ -0,0 +1,65 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package push
+
+import (
+	"fmt"
+
+	"github.com/grafana/mimir/pkg/mimirpb"
+)
+
+// supplierFunc should return either a non-nil body or a non-nil error. The cleanup function can be nil.
+type supplierFunc func() (*mimirpb.WriteRequest, func(), error)
```

Comment:
[nit] I would name the return params for clarity.

```suggestion
type supplierFunc func() (req *mimirpb.WriteRequest, cleanup func(), err error)
```